pkg/dedup/merge_iter.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package dedup
+
+import (
+	"math"
+
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+const UseMergedSeries = "use_merged_series"
+
+// mergedSeries is a storage.Series that implements a simple merge sort algorithm.
+// when replicas has conflict values at the same timestamp, the first replica will be selected.
+type mergedSeries struct {
+	lset     labels.Labels
+	replicas []storage.Series
+}
+
+func NewMergedSeries(lset labels.Labels, replicas []storage.Series) storage.Series {
+	return &mergedSeries{
+		lset:     lset,
+		replicas: replicas,
+	}
+}
+
+func (m *mergedSeries) Labels() labels.Labels {
+	return m.lset
+}
+func (m *mergedSeries) Iterator(_ chunkenc.Iterator) chunkenc.Iterator {
+	iters := make([]chunkenc.Iterator, 0, len(m.replicas))
+	oks := make([]bool, 0, len(m.replicas))
+	for _, r := range m.replicas {
+		it := r.Iterator(nil)
+		ok := it.Next() != chunkenc.ValNone // iterate to the first value.
+		iters = append(iters, it)
+		oks = append(oks, ok)
+	}
+	return &mergedSeriesIterator{
+		iters:    iters,
+		oks:      oks,
+		lastT:    math.MinInt64,
+		lastIter: nil, // behavior is undefined if At() is called before Next(), here we panic if it happens.
+	}
+}
+
+type mergedSeriesIterator struct {
+	iters []chunkenc.Iterator
+	oks   []bool
+
+	lastT    int64
+	lastIter chunkenc.Iterator
+}
+
+func (m *mergedSeriesIterator) Next() chunkenc.ValueType {
+	return m.Seek(m.lastT + initialPenalty) // apply penalty to avoid selecting samples too close
+}
+
+func (m *mergedSeriesIterator) Seek(t int64) chunkenc.ValueType {
+	if len(m.iters) == 0 {
+		return chunkenc.ValNone
+	}
+
+	picked := int64(math.MaxInt64)
+	for i, it := range m.iters {
+		if !m.oks[i] {
+			continue
+		}
+		if it == m.lastIter || it.AtT() <= m.lastT {
```

Comment:
Should this if condition be `it == m.lastIter || it.AtT() < t`? 
It's possible that all iter in `m.iters` has `AtT() < t` (`ok` is true) and `AtT > m.lastT`. In that case, all of them need to seek to pass `t`.

Reply:
we want to move in advance if it currently at the sample timestamp as last one, this test case will fail if not doing so: 

```
Multi dedup labels - data points absent
```

Reply:
I think the current implementation fails the following test case. (though in practice this case may not happen at all, but this mergedSeries implementation doesn't claim any assumption for the input time series).

```
itr_1:  [{10000, -1}]
itr_2: [{11000, -2}]
==== After calling Next() --> m.Seek(math.MinInt64 + 5000) ====
lastIter = itr_1
lastT = 10000
itr_1 = {10000, -1}
itr_2 = {11000, -2}

==== After calling Next() --> m.Seek(10000 + 5000) ====
itr_1 = ValNone
itr_2 = {11000, -2}

itr_2.AtT() < 15000 (a.k.a. the argument t in Seek(t))

No iter is picked, a.k.a. picked == math.MaxInt64

```

Reply:
Oh, that's intended, we don't want high frequent sampling, so the penalty is 5000 which stands for 5 seconds.